### ISSUE TITLE
Precompile error when using Plots.jl

### ISSUE THREAD

**Issue by loorke** [tbreloff/Plots.jl#545]
I got fresh Julia recently. Now I have trouble with Plots.jl.

```
julia> using Plots
INFO: Precompiling module Plots.
WARNING: Base.drop is deprecated.
  likely near /home/loorke/.julia/v0.6/Showoff/src/Showoff.jl:6
WARNING: The call to compilecache failed to create a usable precompiled cache file for module Showoff. Got:
WARNING: Module Iterators uuid did not match cache file.
WARNING: eval from module Main to Plots:    
Expr(:call, Expr(:., :Base, :include_from_node1)::Any, "/home/loorke/.julia/v0.6/Showoff/src/Showoff.jl")::Any
  ** incremental compilation may be broken for this module **

WARNING: imported binding for Iterators overwritten in module Main
WARNING: Base.drop is deprecated.
  likely near /home/loorke/.julia/v0.6/Showoff/src/Showoff.jl:6
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:237
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:238
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:240
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:241
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:242
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:244
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:245
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:246
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:248
WARNING: Base.cycle is deprecated.
  likely near /home/loorke/.julia/v0.6/Plots/src/utils.jl:249
WARNING: Module Iterators uuid did not match cache file.
ERROR: Module Plots declares __precompile__(true) but require failed to create a usable precompiled cache file.
 in require(::Symbol) at ./loading.jl:458
```

My `versioninfo()`:

```
Julia Version 0.6.0-dev.1047
Commit ff0674a (2016-10-20 13:41 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i3-4005U CPU @ 1.70GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)
```

**Comment by SimonDanisch** [comment]
Yeah, this is quite unfortunate, since Base exports Iterator now, which is also a package. I still haven't looked enough into it to figure out what the appropriate fix is for this!

**Comment by ChrisRackauckas** [comment]
Yeah, I got this on v0.6. It seems Plots.jl is no longer working on v0.6 very well. There's a few other issues. I would suggest not using the nightly for serious work anymore: big changes are happening.

**Comment by ChrisRackauckas** [comment]
Try removing precompilation and see if Plots is usable again.

**Comment by tbreloff** [comment]
Precompilation is removed on master: https://github.com/tbreloff/Plots.jl/pull/554

**Comment by dlfivefifty** [comment]
I think `__precompile__()` should be added back and support for 0.6 dropped until it's fixed, as otherwise I'm going to have to put an upper bound on Plots (it's way too slow to compile every package that depends on Plots everytime

**Comment by ChrisRackauckas** [comment]
How much of a difference does it make? Plots does lazy loading of the backends which is the biggest cost in most cases.

**Comment by dlfivefifty** [comment]
The issue is that every package that depends on Plots has to then also be compiled on the fly.  ApproxFun on its own takes 17s to compile.  I'm not going back to the days of Julia 0.3 where everything takes minutes to use!

**Comment by ChrisRackauckas** [comment]
Why does it depend on Plots and not RecipesBase?

**Comment by dlfivefifty** [comment]
Ah because I didn't know about RecipesBase 🙈

**Comment by dlfivefifty** [comment]
But `@userplot` is in Plots.jl

**Comment by ChrisRackauckas** [comment]
You can't use a user recipe? `@userplot` is just shorthand for setting it up. It should probably move to RecipesBase if it can. If there's not an issue open for that, I think that's a good issue.

**Comment by dlfivefifty** [comment]
This seems to be the issue:
https://github.com/JuliaPlots/RecipesBase.jl/issues/12

**Comment by dlfivefifty** [comment]
I'll just comment out the user plots for now, they aren't used that much

**Comment by ChrisRackauckas** [comment]
Yeah, thanks for pointing me there. I think that should be changed. But also, @tbreloff is doing some changes to Plots.jl to make it more of a Base package and moving backends out (I believe?). If that's the case, Plots.jl may actually because the proper Base package that fixes this problem (and the backend loading problem).

**Comment by jtravs** [comment]
But this doesn't change the fact that any package depending on Plots.jl cannot now be precompiled. This is a big issue for me. This is just a temporary measure right?

**Comment by tbreloff** [comment]
1) so far the reorg experiments in Cantina.jl are a big failure. Time to
first plot is still horrible
2) I'm inclined to agree and say screw 0.6 if we can get 0.5 time to first
plot back through precompilation. We can always have a 0.6 branch that
removes it.

**Comment by tbreloff** [comment]
No precompilation on 0.5:

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.1-pre+27 (2016-11-04 08:41 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 25f7066 (70 days old release-0.5)
|__/                   |  x86_64-linux-gnu

julia> tic(); using Plots; @time plot(rand(10))
toc()
 10.560437 seconds (10.99 M allocations: 462.265 MB, 2.24% gc time)

julia> toc()
elapsed time: 23.619057837 seconds
23.619057837
```

Including initial precompilation:

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.1-pre+27 (2016-11-04 08:41 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 25f7066 (70 days old release-0.5)
|__/                   |  x86_64-linux-gnu

julia> tic(); using Plots; @time plot(rand(10))
INFO: Precompiling module Plots.

 11.766984 seconds (12.13 M allocations: 507.195 MB, 3.08% gc time)

julia> toc()
elapsed time: 27.738597754 seconds
27.738597754
```

After including precompilation again:

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.1-pre+27 (2016-11-04 08:41 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 25f7066 (70 days old release-0.5)
|__/                   |  x86_64-linux-gnu

julia> tic(); using Plots; @time plot(rand(10))

 11.119282 seconds (12.05 M allocations: 503.569 MB, 3.34% gc time)

julia> toc()
elapsed time: 17.097527081 seconds
17.097527081
```

tl;dr Precompilation doesn't solve all our startup problems, but certainly helps a little.

I'm suspicious that the whole recipe design is not precompilation-friendly because it involves adding to a method which Plots doesn't "own".  In this case I think those method defs are just ignored in precompilation anyway.

Anyone with more knowledge/experience about precompilation would be a big help... volunteers please!

**Comment by tbreloff** [comment]
I have just created a new `julia0.6` branch which leaves `__precompile__(false)`... it's turned back on for master and dev.

**Comment by dlfivefifty** [comment]
You could use SnoopCompile.jl to determine where the compile time is coming from, and make pull requests to add to the precompiled function lists in other packages.

**Comment by KristofferC** [comment]
I've only seen snoopcompile been useful together with also including the package in the built sysimg.jl.

**Comment by dlfivefifty** [comment]
I definitely see an improvementin ApproxFun.  With `_precompile_()`:
```julia
julia> using ApproxFun

julia> @time Fun()
  0.206041 seconds (373.25 k allocations: 15.432 MB)
Fun(Chebyshev(【-1.0,1.0】),[0.0,1.0])
```
Without `_precompile_()`:
```julia
julia> using ApproxFun
INFO: Recompiling stale cache file /Users/solver/.julia/lib/v0.5/ApproxFun.ji for module ApproxFun.

julia> @time Fun()
  1.093800 seconds (2.24 M allocations: 91.570 MB, 14.13% gc time)
Fun(Chebyshev(【-1.0,1.0】),[0.0,1.0])
```

**Comment by dlfivefifty** [comment]
PS Once compiled:
```julia
julia> @time Fun()
  0.000037 seconds (12 allocations: 624 bytes)
Fun(Chebyshev(【-1.0,1.0】),[0.0,1.0])
```
